tonic-web/src/call.rs

```diff
@@ -160,32 +163,47 @@ where
     B: Body<Data = Bytes>,
     B::Error: Error,
 {
+    // Poll body for data, decoding (e.g. via Base64 if necessary) and returning frames
+    // to the caller. If the caller is a client, it should look for trailers before
+    // returning these frames.
     fn poll_decode(
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-    ) -> Poll<Option<Result<B::Data, Status>>> {
+    ) -> Poll<Option<Result<Frame<B::Data>, Status>>> {
         match self.encoding {
             Encoding::Base64 => loop {
                 if let Some(bytes) = self.as_mut().decode_chunk()? {
-                    return Poll::Ready(Some(Ok(bytes)));
+                    return Poll::Ready(Some(Ok(Frame::data(bytes))));
                 }
 
                 let mut this = self.as_mut().project();
 
-                match ready!(this.inner.as_mut().poll_data(cx)) {
-                    Some(Ok(data)) => this.buf.put(data),
+                match ready!(this.inner.as_mut().poll_frame(cx)) {
+                    Some(Ok(frame)) if frame.is_data() => this.buf.put(frame.into_data().unwrap()),
+                    Some(Ok(frame)) if frame.is_trailers() => {
+                        return Poll::Ready(Some(Err(internal_error(
+                            "malformed base64 request has unencoded trailers",
+                        ))))
+                    }
+                    Some(Ok(_)) => {
+                        unreachable!("unexpected frame type")
```

Comment:
Frame is internally an enum with only Data and Trailers - so this should be unreachable. But I guess to future-proof against http_body API changes, this can be an internal error.

Reply:
I think returning an error might be better here -- avoids another panic path even though I agree with you it seems actually unreachable in practice.